pkg/collector/tcp.go

```diff
@@ -49,24 +51,23 @@ func (cp *CollectingProcess) startTCPServer() {
 					return
 				}
 			}
+			cp.wg.Add(1)
 			go cp.handleTCPClient(conn, stopCh)
 		}
 	}(cp.stopChan)
 	<-cp.stopChan
+	listener.Close()
 }
 
-func (cp *CollectingProcess) handleTCPClient(conn net.Conn, stopCh chan struct{}) {
+func (cp *CollectingProcess) handleTCPClient(conn net.Conn, stopChan chan struct{}) {
 	address := conn.RemoteAddr().String()
 	client := cp.createClient()
 	cp.addClient(address, client)
-	defer conn.Close()
 	buff := make([]byte, cp.maxBufferSize)
-	for {
-		select {
-		case <-stopCh:
-			cp.deleteClient(address)
-			return
-		default:
+	defer cp.wg.Done()
+	defer conn.Close()
+	go func() {
```

Comment:
After connection is closed, `conn.Read(buff)` will have `io.EOF` error then it will return from the goroutine and finish corresponding waitgroup. Reason we cannot use `for{ select{ case: <-cp.stopChan default:}}` is that each iteration we will enter `conn.Read` and wait for input there, the stop message cannot be caught.